aya/src/symbols.rs

```diff
@@ -0,0 +1,688 @@
+#[derive(Debug, Error)]
+pub enum CreateResolverError {
+    /// Failed to create a kernel symbol resolver
+    #[error("Error creating KernelSymbolResolver: {error}")]
+    KernelSymbolResolverBuildError {
+        #[source]
+        /// The original io::Error
+        error: io::Error,
+    },
+}
+
+/// Stores all information that was resolved for a specific symbol
+#[derive(Debug, Clone)]
+pub struct SymbolInfo {
+    /// Virtual memory address of a certain function
+    virtual_address: u64,
+    /// Address of a function inside an object. `None` if it was not resolved
+    object_address: Option<u64>,
+    /// The PID of the process in user space
+    /// `None` is the symbol if the address is a kernel function
+    process_id: Option<u32>,
+    /// The resolved function name
+    function_name: Option<String>,
+    /// Path to the object which defines the symbol
+    object_path: Option<PathBuf>,
+}
+
+impl SymbolInfo {
+    /// Creates a `SymbolInfo` instance for a kernel-space function which hasn't been resolved yet
+    pub fn unresolved_kernel(address: u64) -> Self {
+        Self {
+            virtual_address: address,
+            object_address: Some(address),
+            function_name: None,
+            object_path: None,
+            process_id: None,
+        }
+    }
+
+    /// Creates a `SymbolInfo` instance for a user-space function which hasn't been resolved yet
+    pub fn unresolved_user(process_id: u32, address: u64) -> Self {
+        Self {
+            virtual_address: address,
+            object_address: None,
+            function_name: None,
+            object_path: None,
+            process_id: Some(process_id),
+        }
+    }
+
+    /// Whether the symbol refers to a kernel-space function
+    pub fn is_kernel(&self) -> bool {
+        self.process_id.is_none()
+    }
+
+    /// Whether the symbol refers to a user-space function
+    pub fn is_user(&self) -> bool {
+        self.process_id.is_some()
+    }
+
+    /// Returns the ID of the process for which the symbol was recorded
+    pub fn process_id(&self) -> Option<u32> {
+        self.process_id
+    }
+
+    /// Virtual memory address of a certain function
+    pub fn virtual_address(&self) -> u64 {
+        self.virtual_address
+    }
+
+    /// Address of a function inside an object. `None` if it was not resolved
+    pub fn object_address(&self) -> Option<u64> {
+        self.object_address
+    }
+
+    /// Path to the object which defines the symbol
+    pub fn object_path(&self) -> Option<&Path> {
+        self.object_path.as_deref()
+    }
+
+    /// The resolved function name
+    pub fn function_name(&self) -> Option<&str> {
+        self.function_name.as_deref()
+    }
+}
+
+/// Resolves a symbol based on it's address.
+///
+/// See [`DefaultResolver`] for exemplaric usage.
+pub trait SymbolResolver {
+    /// Resolves a symbol based on it's address
+    fn resolve(&self, symbol: &mut SymbolInfo);
```

Comment:
I chose this API with the intention of "you provide all the info you have, and the resolver fetches the rest". Alternatives could be:

```
fn resolve(&self, symbol: &SymbolInfo) -> Option<Symbolnfo>;
```
or
```
fn resolve(&self, symbol: &SymbolInfo) -> Vec<Symbolnfo>;
```
which don't mutate the initial info. But at the the former is also achievable with the current API, and cloning the `sybol` before.

Reply:
I find this a bit confusing, wouldn't it be clearer if we had something like:

```

let user_resolver = FileResolver::new("/some/path")?;
let sym = user_resolver.resolve(123);

let user_resolver = PidResolver::new(42)?;
let sym = user_resolver.resolve(123);

let kernel_resolver = KernelResolver()?;
let sym = kernel_resolver.resolve(123)
```

Reply:
See the comment at https://github.com/aya-rs/aya/pull/339/files#r933884841 on what could be public. That could certainly also be done, but it means there's no common trait (and thereby also builtin mechnaism for composition/caching/etc) anymore. Also the question what `sym` contains in those results. Is it always the same, or an individual result type per resolver?

In your example, I guess `PidResolver` will not resolve a symbol but just translate the address? It is `ProcMemMap`?